Raycaster in ThreeJS.jl

izaid opened the pull request:
This PR adds a simple raycaster to ThreeJS.jl. It triggers on an arbitrary JavaScript event, and uses machinery from Escher to callback into Julia as a `Signal`. I think this goes a long way towards adding more complex interactivity. In the future, we can extend this to Jupyter as well.

There will be a follow-on PR which provides a way to add more complex behavior directly into the JavaScript, as well as an example demonstrating that.

rohitvarkey commented on assets/bower_components/three-js/three-js.html:
```diff
@@ -1418,3 +1418,49 @@
+
+                var intersects = that.object.intersectObjects(scene.children);
+                if (intersects.length > 0) {
+                    Escher.send({command: 'signal-update', data: {signalId: that.signal, value: intersects[0].point}});
```
Having `Escher` here would throw up an error when this is loaded in any other context. Can you add a check to check if `Escher` is defined and only then use this JS?

shashi replied:
Using `Escher.send` directly this way is not how it's done everywhere in Escher.

How it would work in Escher is, you'd fire `sugnal-update` _event_ instead. Then `subscribe` will automatically set up the signal from that element's `signal-update` events... Instead of the way you do it in the `raycaster` function. So instead of `raycaster(signal, options)` you'd have `subscribe(signal, raycaster(options))`... You can also set up a way to "interpret" the data coming in from the browser like [this](https://github.com/shashi/Escher.jl/blob/4dd34f36dd846fe9cc7abafda8d712541b7fd45c/src/basics/behavior.jl#L85-L118)

I admit this could be simpler though - but I need to work on that. I'm okay with getting this in if it serves your purpose, provided it does not stop this package from working in Jupyter.

izaid replied:
@shashi I'd much rather do things the way you describe, but is there an example I can follow to make this work? I think the issue is it's unclear to me how that should work -- is there some sample JavaScript somewhere in Escher I can look at that does this?

izaid replied:
@shashi Also, if I try and do `subscribe(signal, raycaster)`, I get the following error:

```
MethodError: `subscribe` has no method matching subscribe(::Reactive.Signal{Dict{Any,Any}}, ::Patchwork.Elem{:xhtml,symbol("three-js-raycaster")})
Closest candidates are:
  subscribe(::Any, ::Any, !Matched::Any)
  subscribe(::Reactive.Signal{T}, !Matched::Escher.WithIntent)
  subscribe(::Reactive.Signal{T}, !Matched::Escher.OuterListener{T<:Union{Escher.Collector,Escher.Sampler}})
```

shashi replied:
Okay, so what you're trying to add would be called a "behavior" in Escher. (Note that it's not the same as behavior in polymer - Escher used the word before Polymer 1.0 first did), a `Behavior` is a subtype of `Tile`, but it can "give out a signal".

Let's consider the clickable behavior -- it takes a tile and makes it "clickable" - you can then listen to the click on the input tile. For example.
```julia
el=plaintext("Hello, World!")
clicks = Signal{Escher.MouseButton}(leftbutton)
el_subscribed = subscribe(clicks, clickable(el, buttons=[leftbutton, rightbutton]))
```

Now clicking on `el_subscribed` will update the `clicks` signal.

Here's how this works:

1) The output is something like this: 
```
    <div>Hello, World!
        <clickable-behavior buttons=[1, 3]>
           <signal-transport .../>
         </clickable-behavior>
</div>
```
So you can imagine that `clickable` added the `<clickable-behavior>` tag to the plain text, and `subscribe` added `signal-transport` to _that_ tag. See their respective "render" functions ([clickable](https://github.com/shashi/Escher.jl/blob/master/src/basics/behavior.jl#L112-L118) & [subscribe](https://github.com/shashi/Escher.jl/blob/master/src/basics/signal.jl#L339-L346).

2) When <clickable-behavior> element is loaded, it [adds a "click" listener](https://github.com/shashi/Escher.jl/blob/master/assets/behavior.html#L82-L90) to the parent element, in this handler, it detects if the left or the right mouse button was fired, then calls [`Mixins.ReactiveSignal.update`](https://github.com/shashi/Escher.jl/blob/master/assets/signals.html#L51) with `1` or `3` indicating left or right button - (2 is scroll button)

3) as you  probably noticed, the `Subscription` tile type has an`intent` field which is used together with the signal object. This is an object that tells Julia how to interpret updates coming in from the tile being subscribed to - in our case the `Clickable` tile. When an intent is not specified in `subscribe`, [a default intent is used](https://github.com/shashi/Escher.jl/blob/master/src/basics/signal.jl#L307-L308) (simply the return value of `default_intent` on the input tile). The default intent of a `Clickable` is `ClickIntent`. You may have also noticed that the numbers `1` and `3` are converted to `leftbutton` ==LeftButton() and `rightbutton` ==RightButton() objects. This is the purpose of the `ClickIntent`.

When a message is received from a behavior tile such as clickable-behavior, Escher calls the `interpret` function with the intent of the behavior and the message. In our case `interpret(ClickIntent(), 1)` or `interpret(ClickIntent(),2)` which [calls this](https://github.com/shashi/Escher.jl/blob/master/src/basics/behavior.jl#L105-L110) which returns the value that finally gets pushed to the signal that subscribed to the behavior.

4) There is one last optional piece to this dance: `wrapbehavior`. This is used to associate default behaviors to _Tiles_, for example, a [Button has a default behavior of `clickable`](https://github.com/shashi/Escher.jl/blob/master/src/library/widgets.jl#L60-L61). This makes it possible to say, `subscribe(clicks, button())` which becomes `subscribe(clicks, wrapbehavior(button()))` (but not `subscribe(clicks, "hi")` since "hi" doesn't have a wrapbehavior method.) I don't think wrapbehavior is useful for you.

Coming to this PR:

1. You will need to define a type `immutable Raycaster <: Behavior; options; tile end` which has all the configuration options you need. Then add a method to Escher.render to render it, something like `render(r::Raycaster, state) = render(r.tile, state) << Elem(:three-js-raycaster)`
2. Add the `EscherMixins.ReactiveSignal` polymer behavior to the three js raycaster element, then call `elem = this.findElem(); this.update(elem, {my data...})`
3. Create `RaycasterIntent` which can read `{my data...}` coming from JS to create Julia types representing the update. Also, you will need to define, `Escher.default_intent(x::Raycaster) = RaycasterIntent()`.

This should be enough.

Phew, I hope that is mostly clear. I was always afraid of explaining this because I know I'd definitely fail at it, shoot any questions you have.

It seems that ThreeJS is only dependent on Patchwork and not on Escher. But this PR needs both, so you can probably put these definitions in `@require Escher begin ... end`

izaid replied:
@shashi Thanks for that really great write-up! I'm really excited to see if we can get this working.

I've just updated the PR, following your steps exactly (I think). Almost everything appears to work, but I'm failing at the very last part on the Julia side. I'm trying to append the subscribed raycaster (i.e., the output of `subscribe(my_signal, Raycaster()`) to the virtual DOM as per `ThreeJS.camera(...) << subscribed_raycaster`. I get the error:

``` 
MethodError: `<<` has no method matching <<(::Patchwork.Elem{:xhtml,:div}, ::Escher.Subscription)
```

To me, this looks like an issue with appending. What do you think?

rohitvarkey commented:
@shashi Do you mind having a look over at the Escher JavaScript stuff here? I'm not totally familiar with that.

shashi commented on src/render.jl:
```diff
@@ -614,3 +614,14 @@ function text(x::Float64, y::Float64, z::Float64, content::AbstractString)
+    Create a raycaster that triggers on the JavaScript event `event`.
+    """
+    function raycaster(signal::Signal, event::AbstractString = "")
+        Elem(:"three-js-raycaster"; :signal => Escher.makeid((signal, Escher.ToType{Dict{UTF8String, Float64}}())), :event => event)
```
this is what I was referring to in the other comment, `makeid` is an implementation detail...

izaid commented:
@rohitvarkey I updated this with Shashi's advice. It should now be ready to go. There is a corresponding PR on the JavaScript at https://github.com/rohitvarkey/three-js/pull/5. I've also added the Julia package `Requires` as a dependency, to handle the `Escher` stuff.

I'm pretty happy with this, we should be good to go!

shashi commented:
Looks good to me! 👍

izaid commented:
@shashi Yeah, it's a `Dict` with `x`, `y`, and `z` keys as `Float64`. Maybe we should make it a point, but I don't feel strongly atm. I'd just leave it for now.

rohitvarkey commented:
@izaid Awesome stuff! The examples are really cool. 😄 

@shashi Can you tag new releases of Escher and Patchwork, so that I can tag a release for ThreeJS?

shashi commented:
Done and done!

rohitvarkey commented:
ThreeJS `v0.3.0` tagged and released too. 😄
